File: ITAM/django-backend/config/permissions.py
"""Role-based permissions for DRF.

Three roles:
1. ADMIN - Full system access (users, settings, all CRUD)
2. IT_TEAM - Operational access (assets, assignments, maintenance, etc.)
3. FINANCE - Read-only access to reports and underlying data for exports
"""

from rest_framework.permissions import BasePermission, SAFE_METHODS

from apps.users.models import UserRole


ADMIN_ROLES = {UserRole.ADMIN}
IT_ROLES = {UserRole.ADMIN, UserRole.IT_TEAM}
REPORT_ROLES = {UserRole.ADMIN, UserRole.IT_TEAM, UserRole.FINANCE}


def _user_role(request) -> str | None:
    if not request.user or not request.user.is_authenticated:
        return None
    return getattr(request.user, "role", None)


class IsAdmin(BasePermission):
    """Only Admin."""

    def has_permission(self, request, view):
        return _user_role(request) in ADMIN_ROLES


class IsITTeamOrAdmin(BasePermission):
    """IT Team or Admin."""

    def has_permission(self, request, view):
        return _user_role(request) in IT_ROLES


class CanAccessReports(BasePermission):
    """Admin, IT Team, or Finance — for report generation and read access."""

    def has_permission(self, request, view):
        return _user_role(request) in REPORT_ROLES


class IsITTeamOrAdminOrReadOnly(BasePermission):
    """GET/HEAD/OPTIONS for Admin, IT Team, and Finance; writes for Admin and IT Team."""

    def has_permission(self, request, view):
        role = _user_role(request)
        if role is None:
            return False
        if request.method in SAFE_METHODS:
            return role in REPORT_ROLES
        return role in IT_ROLES


class IsSelfOrITTeamOrAdmin(BasePermission):
    """Allow a user to edit their own record; IT Team/Admin can edit anyone."""

    def has_object_permission(self, request, view, obj):
        role = _user_role(request)
        if role is None:
            return False
        if obj.pk == request.user.pk:
            return True
        return role in IT_ROLES


class IsFinanceReadOnly(BasePermission):
    """Finance users may only use safe HTTP methods."""

    def has_permission(self, request, view):
        role = _user_role(request)
        if role is None:
            return False
        if role == UserRole.FINANCE:
            return request.method in SAFE_METHODS
        return True


# Backward-compatible aliases used by audits/attachments modules
IsITStaffOrAdmin = IsITTeamOrAdmin
IsITStaffOrAdminOrReadOnly = IsITTeamOrAdminOrReadOnly
IsSelfOrITStaffOrAdmin = IsSelfOrITTeamOrAdmin
IsITAdmin = IsITTeamOrAdmin
IsAssetManager = IsITTeamOrAdmin
IsDepartmentManager = IsITTeamOrAdmin
IsAuditor = IsITTeamOrAdmin
IsAssetManagerOrReadOnly = IsITTeamOrAdminOrReadOnly

File: ITAM/frontend/src/lib/permissions.ts
import { UserRole } from '@/lib/authRole';

/** Paths accessible per role */
const ROLE_PATHS: Record<UserRole, Set<string>> = {
  admin: new Set([
    '/',
    '/assets',
    '/locations',
    '/assignments',
    '/maintenance',
    '/software',
    '/audits',
    '/reports',
    '/manufacturers',
    '/notifications',
    '/users',
    '/settings',
    '/profile',
  ]),
  it_team: new Set([
    '/',
    '/assets',
    '/locations',
    '/assignments',
    '/maintenance',
    '/software',
    '/audits',
    '/reports',
    '/manufacturers',
    '/notifications',
    '/profile',
  ]),
  finance: new Set(['/reports', '/profile']),
};

export function canAccessPath(role: UserRole, path: string): boolean {
  const allowed = ROLE_PATHS[role];
  if (!allowed) return false;
  if (allowed.has(path)) return true;
  // Block finance from all other routes
  return false;
}

export function getDefaultPath(role: UserRole): string {
  if (role === 'finance') return '/reports';
  return '/';
}

export function canWrite(role: UserRole): boolean {
  return role === 'admin' || role === 'it_team';
}

export function canDelete(role: UserRole): boolean {
  return role === 'admin';
}

export function isAdmin(role: UserRole): boolean {
  return role === 'admin';
}

export function isITTeam(role: UserRole): boolean {
  return role === 'it_team';
}

export function isFinance(role: UserRole): boolean {
  return role === 'finance';
}

export function getRoleLabel(role: UserRole): string {
  switch (role) {
    case 'admin':
      return 'Admin';
    case 'it_team':
      return 'IT Team';
    case 'finance':
      return 'Finance';
    default:
      return role;
  }
}
